src/enumerable.cr

```diff
@@ -609,6 +609,37 @@ module Enumerable(T)
     memo
   end
 
+  # Similar to `reduce`, but instead of raising when the input is empty,
+  # return `nil`
+  #
+  # ```
+  # ([] of Int32).reduce? { |acc, i| acc + i } # => nil
+  # ```
+  def reduce?
+    memo = uninitialized T
```

Comment:
I'm not certain on how to do this - most of the other `private def` in enumerable take a block as an argument. Can you advise what this would look like?

Also, to keep things consistent with this file, the name should be `private def reduce_internal` and the return value `{found, memo}`.